.github/actions/detect-workflow-js/src/detect.ts

```diff
@@ -60,42 +60,55 @@ export async function detectWorkflowFromContext(
     );
   }
 
-  // Filter referenced_workflows for slsa-github-generator repositories,
-  // on any fork.
-  // TODO(https://github.com/actions/runner/issues/2417): When
-  // GITHUB_JOB_WORKFLOW_SHA becomes fully functional, the OIDC token
-  // detection can be removed and we can identify the current reusable workflow
-  // through the sha of a referenced workflow, fully supporting all triggers
-  // without the repository filter.
   let [repository, ref, workflow] = ["", "", ""];
-  for (const reusableWorkflow of workflowData.referenced_workflows) {
-    const workflowPath = reusableWorkflow.path.split("@", 1);
-    const [workflowOwner, workflowRepo, ...workflowArray] =
-      workflowPath[0].split("/");
-    const tmpRepository = [workflowOwner, workflowRepo].join("/");
-    if (!reusableWorkflow.ref) {
-      return Promise.reject(
-        Error(
-          "Referenced workflow missing ref: was the workflow invoked by digest?"
-        )
-      );
-    }
-    const tmpRef = reusableWorkflow.ref;
-    const tmpWorkflow = workflowArray.join("/");
-    if (workflowRepo === "slsa-github-generator") {
-      // If there are multiple invocations of reusable workflows in
-      // a single caller workflow, ensure that the repositories and refs are
-      // the same.
-      if (repository !== "" && repository !== tmpRepository) {
-        return Promise.reject(Error("Unexpected mismatch of repositories"));
+
+  // If this is a slsa-github-generator repository or fork, then look
+  // for the repo and head SHA from the pull_request event value.
+  if (
+    workflowData.event === "pull_request" &&
+    workflowData.repository.name === "slsa-github-generator"
+  ) {
+    ref = workflowData.head_sha;
+    repository = workflowData.head_repository.full_name;
+    workflow = workflowData.path;
+  } else {
+    // Otherwise this is an external repository.
+    // Filter referenced_workflows for slsa-github-generator repositories.
+    // TODO(https://github.com/actions/runner/issues/2417): When
+    // GITHUB_JOB_WORKFLOW_SHA becomes fully functional, the OIDC token
+    // detection can be removed and we can identify the current reusable workflow
+    // through the sha of a referenced workflow, fully supporting all triggers
+    // without the repository filter.
+    for (const reusableWorkflow of workflowData.referenced_workflows) {
+      const workflowPath = reusableWorkflow.path.split("@", 1);
+      const [workflowOwner, workflowRepo, ...workflowArray] =
+        workflowPath[0].split("/");
+      const tmpRepository = [workflowOwner, workflowRepo].join("/");
+      if (!reusableWorkflow.ref) {
```

Comment:
This seems like it would fail if any reusable workflows in the chain are referenced by digest when I think we only care about the slsa-github-generator ones? Perhaps it should be this?

```suggestion
      if (workflowRepo === "slsa-github-generator" && !reusableWorkflow.ref) {
```